src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributeList.cs

```diff
@@ -2,73 +2,52 @@
     /// </typeparam>
-    public class ReadOnlyTagHelperAttributeList<TAttribute> : IReadOnlyList<TAttribute>
-        where TAttribute : IReadOnlyTagHelperAttribute
+    public class ReadOnlyTagHelperAttributeList : ReadOnlyCollection<TagHelperAttribute>
     {
-        private static readonly IReadOnlyList<TAttribute> EmptyList =
-#if NET451
-            new TAttribute[0];
-#else
-            Array.Empty<TAttribute>();
-#endif
+        private static readonly IReadOnlyList<TagHelperAttribute> EmptyList = new TagHelperAttribute[0];
 
         /// <summary>
-        /// Instantiates a new instance of <see cref="ReadOnlyTagHelperAttributeList{TAttribute}"/> with an empty
+        /// Instantiates a new instance of <see cref="ReadOnlyTagHelperAttributeList"/> with an empty
         /// collection.
         /// </summary>
         protected ReadOnlyTagHelperAttributeList()
+            : base(new List<TagHelperAttribute>())
         {
-            Attributes = new List<TAttribute>();
         }
 
         /// <summary>
-        /// Instantiates a new instance of <see cref="ReadOnlyTagHelperAttributeList{TAttribute}"/> with the specified
+        /// Instantiates a new instance of <see cref="ReadOnlyTagHelperAttributeList"/> with the specified
         /// <paramref name="attributes"/>.
         /// </summary>
         /// <param name="attributes">The collection to wrap.</param>
-        public ReadOnlyTagHelperAttributeList(IEnumerable<TAttribute> attributes)
+        public ReadOnlyTagHelperAttributeList(IEnumerable<TagHelperAttribute> attributes)
+            : base(new List<TagHelperAttribute>(attributes))
         {
-            if (attributes == null)
-            {
-                throw new ArgumentNullException(nameof(attributes));
-            }
-
-            Attributes = new List<TAttribute>(attributes);
         }
 
         /// <summary>
-        /// The underlying collection of <typeparamref name="TAttribute"/>s.
-        /// </summary>
-        /// <remarks>Intended for use in a non-read-only subclass. Changes to this <see cref="List{TAttribute}"/> will
-        /// affect all getters that <see cref="ReadOnlyTagHelperAttributeList{TAttribute}"/> provides.</remarks>
-        protected List<TAttribute> Attributes { get; }
-
-        /// <inheritdoc />
-        public TAttribute this[int index] => Attributes[index];
-
-        /// <summary>
-        /// Gets the first <typeparamref name="TAttribute"/> with <see cref="IReadOnlyTagHelperAttribute.Name"/>
+        /// Gets the first <typeparamref name="TagHelperAttribute"/> with <see cref="TagHelperAttribute.Name"/>
         /// matching <paramref name="name"/>.
         /// </summary>
         /// <param name="name">
-        /// The <see cref="IReadOnlyTagHelperAttribute.Name"/> of the <typeparamref name="TAttribute"/> to get.
+        /// The <see cref="TagHelperAttribute.Name"/> of the <typeparamref name="TagHelperAttribute"/> to get.
```

Comment:
`<see cref="TagHelperAttribute"/>`
Should have no `<typeparamref/>` elements in this class.